#34: Migration script: execute only if there is a valid license

slauriere opened the pull request:
- Move the migration script call to a block executed only if there is a valid license
- Move the migration macro call to Diagram.WebHome so that Diagram.MigrationScriptMacros contains macro definition only
- Change slightly the formatting

mflorea commented on src/main/resources/Diagram/WebHome.xml:
```diff
@@ -66,6 +65,8 @@
 {{velocity}}
 #set ($diagramClassReference = $services.model.createDocumentReference('', 'Diagram', 'DiagramClass'))
 #if ($services.licensing.licensor.hasLicensureForEntity($diagramClassReference))
+  #includeMacros('Diagram.MigrationScriptMacros')
```
I have some doubts about this old Velocity macro. Using the ``{{include}}`` wiki syntax macro would have been cleaner, but it requires a separate Velocity block so it complicates the code. I hope this old ``includeMacros`` doesn't have hidden issues.

vmassol replied:
+1 to not use this deprecated way. What's the problem with:

```
{{velocity}}
#set ($diagramClassReference = $services.model.createDocumentReference('', 'Diagram', 'DiagramClass'))
#if ($services.licensing.licensor.hasLicensureForEntity($diagramClassReference))
  {{include reference='Diagram.MigrationScriptMacros'/}}
```

?

mflorea replied:
@vmassol as I said, you need to do the include in a separate Velocity block otherwise the macro call ``#selectDiagramsRequiringMigration`` (from the next line) is evaluated before the wiki syntax (include macro) is executed and the macro definition is not found.

vmassol replied:
ok, was just looking at the diff. There's a real question as to whether we want to keep velocity macros for include or not. Right now they're marked deprecated on their extension page, see https://extensions.xwiki.org/xwiki/bin/view/Extension/Include%20Form%20Macro

Right now we're in a grey area: we say they're deprecated but we also consider (see your comments above) that it's ok to use them because they serve some use case better than the rendering macro. 

I think we need a thread on the list to clarify this.

slauriere replied:
I had doubts as well with this macro but since it was not marked as deprecated in 'macros.vm', I thought it might still be acceptable (I had not checked the documentation page that you mention Vincent).

I pushed a [new version](https://github.com/xwikisas/application-diagram/pull/35/commits/97b382bfea427b2ebef92de8b7c4fb419b52da45) with the following changes:
- First of all it fixes an issue with the migration script itself (in the previous commit it was not working anymore, sorry for not testing further beforehand).
- The migration macros are now always included, and executed only if a valid license is available.
- Minor formatting and wording changes were brought.

Note that the migration utility in this version works only for an upgrade of the pro version or an upgrade to it.

Please let me know if the latest commit can be merged or if you notice any issue including formatting ones.

mflorea commented:
@slauriere I'm also wondering if it wouldn't have been easier to simply put the

    {{include reference="Diagram.MigrationScriptMacros"/}}

in a separate Velocity block, under the same condition

    #if ($services.licensing.licensor.hasLicensureForEntity($diagramClassReference))

Yes, the condition would be duplicated in the second Velocity block, but that's acceptable IMO. Do you have time to work on this, otherwise I can ask @oanalavinia or @arcilli to do it.

slauriere commented:
Hi Marius, I can let you know on Monday if I can look into this next week.

slauriere commented:
Marius, We have allocated time for me to investigate this week the comments you raised, on Wednesday or Thursday, hoping it's ok, I'll let you know then – cc @oanalavinia, @arcilli

mflorea commented:
Great, thanks @slauriere.

slauriere commented:
Marius, I confirm the issue was still present since access to MigrationScriptMacros was denied without a license. I fixed that and brought the following improvements:
- Moved `{{include reference="Diagram.MigrationScriptMacros"/}}` to a separate Velocity block as you suggested.
- Wrote the license check as a single (long) line: `#if ($services.licensing.licensor.hasLicensureForEntity($services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))`
- Removed the explicit license check from MigrationScript and MigrationScriptLiveTableResults since these pages are already not viewable in case a license is not available.
- Updated the formatting.

mflorea commented on src/main/resources/Diagram/WebHome.xml:
```diff
@@ -40,7 +40,11 @@
   <content>{{include reference="Licenses.Code.VelocityMacros"/}}
-{{include reference="Diagram.MigrationScriptMacros"/}}
+{{velocity}}
+#if ($services.licensing.licensor.hasLicensureForEntity($services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))
```
```suggestion
#if ($services.licensing.licensor.hasLicensureForEntity(
    $services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))
```

mflorea commented on src/main/resources/Diagram/WebHome.xml:
```diff
@@ -64,14 +68,14 @@
 {{velocity}}
-#set ($diagramClassReference = $services.model.createDocumentReference('', 'Diagram', 'DiagramClass'))
-#if ($services.licensing.licensor.hasLicensureForEntity($diagramClassReference))
+#if ($services.licensing.licensor.hasLicensureForEntity($services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))
```
```suggestion
#if ($services.licensing.licensor.hasLicensureForEntity(
    $services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))
```

The best practice is to wrap long lines (>120 chars) in order to avoid horizontal scroll bars when viewing the source code.

mflorea commented on src/main/resources/Diagram/WebHome.xml:
```diff
@@ -64,14 +68,14 @@
+#if ($services.licensing.licensor.hasLicensureForEntity($services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))
+  #selectDiagramsRequiringMigration()
   #if ($toBeMigratedEntries &amp;&amp; $toBeMigratedEntries.size() &gt; 0)
+    #if ($toBeMigratedEntries.size() &gt; 1)#set ($label = 'diagrams require')#else#set ($label = 'diagram requires')#end
```
Using a translation keys like this:

    key={0} {0,choice,1#diagram requires|1<diagrams require} a migration. Please click {1}here{2} to proceed.

is more elegant.

slauriere replied:
Indeed, thanks for the tip.

mflorea commented:
@slauriere will you have time to finish this in the coming days? I'd like to release a new version of the Diagram app and I want to know if I should wait or not. No problem if you don't have time, we'll include this issue in the next version if that is the case. Thanks.

slauriere commented:
Hi Marius, Sure, I'll handle this by tomorrow. Cheers

mflorea commented:
Thanks! I'll wait then.

acotiuga commented on src/main/resources/Diagram/DiagramTranslations.fr.xml:
```diff
@@ -57,19 +57,19 @@ diagram.livetable.doc.location=Emplacement
 
 # Diagram editor
-diagram.editor.saveAsImageAttachmentError=Erreur à l'enregistrement du diagramme comme un attachement.</content>
+diagram.editor.saveAsImageAttachmentError=Erreur à l'enregistrement du diagramme comme une attachement.</content>
```
Are you sure about this? Isn't `un attachment` or `une piece jointe`?

slauriere replied:
You're right Alex, it's just I re-synced this branch with master, hence erasing previous translation fixes wrongly pushed to this branch initially. We can fix the translation issues either in master or in a dedicated branch, not in this one, right?

acotiuga replied:
Sure, as you wish! Thanks!

slauriere commented on src/main/resources/Diagram/MigrationScript.xml:
```diff
@@ -34,67 +34,65 @@
   <syntaxId>xwiki/2.1</syntaxId>
   <hidden>true</hidden>
-  <content>{{include reference="Diagram.MigrationScriptMacros"/}}
+  <content>{{include reference="Licenses.Code.VelocityMacros"/}}
```
You're right, it's not needed anymore, I've removed the include in the latest commit.

slauriere commented:
The latest commits take into account the change that you suggested Marius, thanks for the review, let me know if anything else's needed before merging. I'll handle the French labels update directly in master, hoping that's ok.

mflorea commented on src/main/resources/Diagram/WebHome.xml:
```diff
@@ -64,14 +69,17 @@
+    $services.model.createDocumentReference('', 'Diagram', 'DiagramClass')))
+  #selectDiagramsRequiringMigration()
   #if ($toBeMigratedEntries &amp;&amp; $toBeMigratedEntries.size() &gt; 0)
+    #if ($toBeMigratedEntries.size() &gt; 1)#set ($label = 'diagrams require')#else#set ($label = 'diagram requires')#end
```
Leftover?

slauriere replied:
Sorry about that, I removed it indeed.
